Add authentication

maggiemay-afk opened the pull request:
First version of admin authentication. 

- Installed react router
- created /login path
- Added error page component for router errors
- Added wildcard route to redirect any other routes back to the home page
- Added login page component 
- Added 'sign in with email and password' option from Firebase
- This works for users with email/password credentials in Firebase
- Added sign out from firebase

hermanator608 commented on src/login.tsx:
```diff
@@ -0,0 +1,110 @@
+
+  const attemptLogin = async () => {
+
+    signInWithEmailAndPassword(auth, username, password)
+      .then((userCredential: any) => {
+      // Signed in
+      alert("success!");
+      setUsername("");
+      setPassword("");
+      setSignedIn(true);
+
+    })
+    .catch((error: any) => {
+      console.log(error.code);
+      console.log(error.message);
+    });
```
If we go into this catch, can we alert to user with a simple error message? For example if it's a `auth/invalid-credential` can we tell the user that their username/password is invalid?

I attempted to login and can't tell if it's working or not and had to check the debug console to realize I was using the wrong password.

hermanator608 commented on src/login.tsx:
```diff
@@ -0,0 +1,110 @@
+      alert("Sign out Success");
+      setSignedIn(false);
+
+    }).catch((error) => {
+      console.log(error.code);
+      console.log(error.message);
+    })
```
Maybe something similar here if we fail to signout?

hermanator608 commented on src/index.css:
```diff
@@ -70,3 +70,24 @@ select {
+  padding-bottom: 15px;
+}
+
+.login-text-field {
+  --brown-glow-drop-shadow: none;
+
+}
+
+.login-button {
+  padding-top: 25px;
+  --brown-glow-drop-shadow: none;
+
+
+}
```
Two things:
* Can we cleanup the empty lines?
* Instead of having to apply this to all the custom classes, you could just make it so nothing on this admin page has the dropshadow styling. Something like:

```
.myClass h1,
.myClass h3,
.myClass p,
.myClass span,
.myClass button,
.myClass select,
.myClass input {
    --brown-glow-drop-shadow: none;
}
```

hermanator608 replied:
Where .myClass is a class on the top level div for the `login` component

hermanator608 replied:
Or even slightly cleaner:
```
.myClass * {
    --brown-glow-drop-shadow: none;
}
```

hermanator608 reviewed (changes requested):
Great work on setting up react-router and firebase login! Few small requests!

maggiemay-afk commented:
@hermanator608 when you get a chance, can you take a look at the error / check fail I'm getting? 
"ReferenceError: TextDecoder is not defined" 

I'm seeing some similar issues online, but not sure about a solution

hermanator608 commented:
> @hermanator608 when you get a chance, can you take a look at the error / check fail I'm getting? "ReferenceError: TextDecoder is not defined"
> 
> I'm seeing some similar issues online, but not sure about a solution

Discussed this outside of github, but seems to be a weird test environment issue - try this https://stackoverflow.com/questions/68468203/why-am-i-getting-textencoder-is-not-defined-in-jest 

May also run into an issue with the test where we are now using <Routes> but the test renders <App> which does not include the <BrowserRouter>

spencer-zepelin reviewed (commented):
Glad to see some much needed updates being made on this project! (especially after the original maintainer disappeared 😉)

hermanator608 commented:
> Glad to see some much needed updates being made on this project! (especially after the original maintainer disappeared 😉)

Just wait and see what is going to come later this year! 💯

maggiemay-afk commented:
Okay, original error: "ReferenceError: TextDecoder is not defined" and subsequent errors in the test file should be fixed.

- Updated App file by removing our check for a valid user
(This was creating bug where an active user couldn't go to the main Ambiance page, it would redirect them back to Admin)
- Put the check for a valid user in the Login page, fixing bug above

Updated test file:
- Mock out Firebase/auth to test authentication
- Add MemoryRouter to test cases so that routes (react router) work
- Add new tests to check our routes with a valid and invalid user
- Add test-ids to Admin and Login page 

Misc. 
- Removed a log in the mouse styles file. Not part of this PR, but was creating some clutter in the console
- Remove AuthProvider from index file, add it to the App instead.

hermanator608 commented on src/error-page.tsx:
```diff
@@ -0,0 +1,16 @@
+import { useRouteError } from "react-router-dom";
+
+
+export default function ErrorPage() {
```
This doesnt look like its used anymore can we delete this?

hermanator608 commented on src/theme.ts:
```diff
@@ -29,10 +29,12 @@ export const theme = createTheme({
-        focused: {
-          borderColor: common.white
+        root: {
+          "&.Mui-focused": {
+            "borderColor": "#fff"
+          }
         }
-      }
+      },
```
Is this necessary? What is it used for? Can we remove if not used?

hermanator608 reviewed (commented):
Great work! Few small questions

hermanator608 reviewed (approved):
LGTM
